---
title: "scUnifrac Report"
header-includes:
   - \usepackage{float}
output: 
  pdf_document:
    toc: true
    number_sections: true
params:
  data: plotData
  outFile: outputFile
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)

knitr::opts_chunk$set(echo = TRUE, autodep=TRUE, cache.comments=FALSE, message=FALSE, warning=FALSE, dpi=300)

addLinkTag<-function(text,link) {
  result<-paste0("<a href='",link,"' target='_blank'>",text,"</a>")
  return(result)
}

figRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("figcap.prefix"), 
           sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      return(caption)
    } else {
      used[label] <<- TRUE
      return(tag[label])
    }
  }
})
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")

tabRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      return(caption)
    } else {
      used[label] <<- TRUE
      return(tag[label])
    }
  }
})

getHeatmapHeight<-function(hdata){
  max(10, nrow(hdata) / 10)
}

options(knitr.table.format = "latex")
```

```{r data, include=FALSE, cache=FALSE}
plotData<-params$data
outputFile<-params$outFile
DISPLAY_FIGURE<-TRUE
```

# Overview

**scUnifrac is to quantify cell subpopulation diversity between two single-cell transcriptome profiles**, which calculates the distance, estimates the statistical significance of the distance, identifies the subpopulations that are significant different between two profiles, finds genes that mark subpopulations, and predicts cell types of subpopulations. If two single-cell RNA-seq profiles have identical cell subpopulation structures (the same subpopulations with the same proportion), the distance is zero, whereas, the distance is one if completely different subpopulations.

  * _**Results Summary **_ - The distance, the statistical significance (p-value), the subpopulations that are different in two profiles, and phylogenetic tree plot and tSNE plot to show the subpopulation structures. 
  * _**Subpopulation view**_ - heatmap of gene expression that mark the statistical significant subpopulations, and predicted cell types of these subpopulations based on Mouse Cell Atlas [Han et al., 2018, Cell, 1091-1107] if scRNA-seq are derived from mouse samples.

# Results Summary

Subpopulation structure:

```{r tableCell, echo=FALSE, results="asis"}
count.table<-plotData$count.table
count.table<-data.frame(count.table, check.names = F)
count.table<-cbind(data.frame(Subpopulation=plotData$samplenames), count.table)
kable(count.table, caption=tabRef("tableCell", "The number of cells in each subpopulation in two scRNA-seq profiles"), row.names=FALSE, align="r")
```

<hr>

```{r tableDiff, echo=FALSE, results="asis"}
df<-t(data.frame("R1"=c("Proportion difference", sprintf("%.2f", plotData$diffval))))
colnames(df)<-c("Subpopulation", plotData$diffnodes)
kable(df, caption=tabRef("tableDiff", "Significantly different subpopulations and their proportion difference between two scRNA-seq profiles"), row.names=FALSE, align="r")
```


<hr>

```{r plotTree, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotTree", "Phylogenetic tree plot of subpopulation structures. Cell subpopulations that are enriched in the first scRNA-seq profile labeled as red, while subpopulations enriched in the second profile are labeled as blue. The line width is proportion to the enrichment. Thicker lines suggest more enriched in one profile. Cell subpopulations that are similar in both samples are labeled as black."), fig.height=8, fig.width=8, fig.pos="H"}
Plotdiff_tree(plotData$count.table,plotData$tree,plotData$diffpop_perm_max,plotData$relpop_perm_max,plotData$legendtxt)
```

<hr>

```{r plotTSNE, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotTSNE", "tSNE plot of subpopulation structures. Each subpopulation that are different between two profiles are denoted as different colors, and subpopulations that are similar between two profiles are denotes as gray. Cells from the first profile are denoted by circle, while cell from the second profile are denoted by star. "), fig.height=8, fig.width=8, fig.pos="H"}
Plotdiff_tSNE(plotData$count.table, plotData$hvgdata_pca, plotData$memb, plotData$diffnodes,plotData$colcluster,plotData$legendtxt)
```

<hr>

```{r dataPreparation, echo=FALSE, error=FALSE, message=FALSE, warning=FALSE}
subRmd<-paste0(outputFile, "_subpopulation.rmd")
fileContent<-c("# Subpopulation View \n")
#for (i in 1:2){
for (i in 1:length(plotData$diffnodes)){
  nodeName<-plotData$diffnodes[i]
  commonContent<-paste0("  nodeName<-plotData$diffnodes[", i, "]\n  clustertext<-paste0(\"Subpopulation \", nodeName, \", \", paste(plotData$count.table[, nodeName], collapse=\":\"))\n")

  fileContent<-c(fileContent, paste0("\\pagebreak\n\n## The cell subpopulation ", nodeName, "\n"))
  
  fileContent<-c(fileContent, paste0("```{r plotMarkergene", nodeName, ", eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align=\"center\", fig.cap=figRef(\"", "plotMarkGene", nodeName, "\", \"Heatmap of gene expressions that mark the cell subpopulation ", nodeName, "\"), fig.height=7, fig.width=7, fig.pos=\"H\"}"))

  fileContent<-c(fileContent, paste0(commonContent,"  Plotdiff_Markergene(plotData$normdata, plotData$memb, nodeName, plotData$colcluster, clustertext)"))
  
  fileContent<-c(fileContent, "```\n")
  
  fileContent<-c(fileContent, paste0("```{r plotCelltype", nodeName, ", eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align=\"center\", fig.cap=figRef(\"", "plotCelltype", nodeName, "\", \"Predicted cell types of the cell subpopulation ", nodeName, "\"), fig.height=7, fig.width=7, fig.pos=\"H\"}"))
  
  fileContent<-c(fileContent, paste0(commonContent, "  Plotdiff_Celltype(plotData$normdata[,plotData$memb==nodeName], plotData$ref.expr, clustertext)"))
  
  fileContent<-c(fileContent, "```\n<hr>\n")
}
writeLines(fileContent, subRmd)
```

```{r child=subRmd, echo=FALSE}
```

```{r cleanup, echo=FALSE, error=FALSE, message=FALSE, warning=FALSE}
file.remove(subRmd)
```

# Session Info

```{r session_info, include=TRUE, echo=TRUE, results='markup'}
devtools::session_info()
```
