%\VignetteIndexEntry{Quantitative assessment of cell population diversity in single-cell landscapes}
%\VignettePackage{scUnifrac}
%\VignetteEngine{knitr::knitr}

\documentclass[12pt]{article}
\usepackage{hyperref}

\author{Quanhu Sheng, Yu Shyr, Qi Liu \\[1em] \small{Center for 
Quantitative Sciences, Vanderbilt University Medical Center, Nashville, USA} \\ 
\small{\texttt{$^*$qi.liu (at) vanderbilt.edu}}}

\title{scUnifrac: a package for quantitative assessment of cell population diversity in single-cell landscapes}

\begin{document}

\maketitle

\begin{abstract}

In this vignette, we demonstrate the application of scUnifrac as a package for quantitative assessment of cell population diversity in single-cell landscapes. 

\vspace{1em}

\textbf{scUnifrac version:} \Sexpr{packageDescription("scUnifrac")$Version}

\vspace{1em}

\begin{center}
    \begin{tabular}{ | l | }
      \hline 
      If you use scUnifrac in published research, please cite:  \\
      \\
      Qi Liu, Charles A. Herring, Quanhu Sheng, et.al,: Quantitative assessment of cell population diversity in single-cell landscapes \\
      \hline 
    \end{tabular}
  \end{center}
\end{abstract}

\tableofcontents

\section{Introduction}

**scUnifrac is to quantify cell subpopulation diversity between two single-cell transcriptome profiles**, which calculates the distance, estimates the statistical significance of the distance, identifies the subpopulations that are significant different between two profiles, finds genes that mark subpopulations, and predicts cell types of subpopulations. If two single-cell RNA-seq profiles have identical cell subpopulation structures (the same subpopulations with the same proportion), the distance is zero, whereas, the distance is one if completely different subpopulations.

  * _**Results Summary **_ - The distance, the statistical significance (p-value), the subpopulations that are different in two profiles, and phylogenetic tree plot and tSNE plot to show the subpopulation structures. 

  * _**Subpopulation view**_ - heatmap of gene expression that mark the statistical significant subpopulations, and predicted cell types of these subpopulations based on Mouse Cell Atlas [Han et al., 2018, Cell, 1091-1107] if scRNA-seq are derived from mouse samples.

\section{Quick start}

Here we show the most basic steps. 

<<quick, echo=TRUE, message=FALSE, warning=FALSE, tidy=FALSE>>=
	library(scUnifrac)
	load(system.file("extdata", "colon1.Rdata", package = "scUnifrac"))
	load(system.file("extdata", "pan1.Rdata", package = "scUnifrac"))
	load(system.file("extdata", "ref.expr.Rdata", package = "scUnifrac"))
	scUnifrac("scUnifracReport.html", colon1, "Colon", pan1, "Pan", ref.expr, genenum=500, ncluster=10, nDim=4, cache=TRUE )
@

\section{Session Info}

<<sessInfo, echo=FALSE, results="asis">>=
toLatex(sessionInfo())
@

\end{document}